In [1]:
import numpy as np
import pandas as pd
import json
import Levenshtein

In [2]:
!pip install python-Levenshtein

You should consider upgrading via the 'pip install --upgrade pip' command.


In [3]:
#df = pd.read_json ('./data/neighbor-districts.json')
f = open('../data/neighbor-districts.json')
neighbor_data = json.load(f)

In [4]:
neighbor_data

{'adilabad/Q15211': ['komram_bheem/Q28170184',
  'nanded/Q692389',
  'yavatmal/Q1804852',
  'nirmal/Q28169750',
  'chandrapur/Q1797274'],
 'agar_malwa/Q15732396': ['rajgarh/Q1833306',
  'shajapur/Q2449803',
  'ratlam/Q2299164',
  'ujjain/Q892641',
  'jhalawar/Q1471417'],
 'agra/Q606343': ['etawah/Q1815288',
  'hathras/Q1814892',
  'dhaulpur/Q1207709',
  'bhind/Q2341700',
  'morena/Q2341467',
  'mathura/Q1773422',
  'etah/Q1773429',
  'firozabad/Q1946950',
  'bharatpur/Q854861'],
 'ahmedabad_district/Q401686': ['anand_district/Q485683',
  'mahesana_district/Q2019694',
  'botad_district/Q14505072',
  'surendranagar_district/Q237535',
  'bhavnagar_district/Q854963',
  'gandhinagar_district/Q1772860',
  'kheda_district/Q1755463'],
 'ahmednagar/Q401744': ['nashik/Q1797269',
  'konkan_division/Q6268840',
  'pune_district/Q1797336',
  'bid/Q814037',
  'osmanabad/Q1647186',
  'aurangabad/Q592942',
  'thane/Q943099',
  'solapur/Q1797263'],
 'aizwal/Q1947322': ['mamit/Q751531',
  'cachar/Q42209'

In [5]:
len(neighbor_data.keys())

723

In [6]:
vacc_data = pd.read_csv('../data/cowin_vaccine_data_districtwise.csv', low_memory=False)
vacc_data

S No State_Code                        State  \
0      NaN        NaN                          NaN   
1      1.0         AN  Andaman and Nicobar Islands   
2      2.0         AN  Andaman and Nicobar Islands   
3      3.0         AN  Andaman and Nicobar Islands   
4      4.0         AP               Andhra Pradesh   
5      5.0         AP               Andhra Pradesh   
6      6.0         AP               Andhra Pradesh   
7      7.0         AP               Andhra Pradesh   
8      8.0         AP               Andhra Pradesh   
9      9.0         AP               Andhra Pradesh   
10    10.0         AP               Andhra Pradesh   
11    11.0         AP               Andhra Pradesh   
12    12.0         AP               Andhra Pradesh   
13    13.0         AP               Andhra Pradesh   
14    14.0         AP               Andhra Pradesh   
15    15.0         AP               Andhra Pradesh   
16    16.0         AP               Andhra Pradesh   
17    17.0         AR            Arunachal Pradesh   
18    18.0         AR            Arunachal Pradesh   
19    19.0         AR            Arunachal Pradesh   
20    20.0         AR            Arunachal Pradesh   
21    21.0         AR            Arunachal Pradesh   
22    22.0         AR            Arunachal Pradesh   
23    23.0         AR            Arunachal Pradesh   
24    24.0         AR            Arunachal Pradesh   
25    25.0         AR            Arunachal Pradesh   
26    26.0         AR            Arunachal Pradesh   
27    27.0         AR            Arunachal Pradesh   
28    28.0         AR            Arunachal Pradesh   
29    29.0         AR            Arunachal Pradesh   
..     ...        ...                          ...   
725  725.0         UT                  Uttarakhand   
726  726.0         UT                  Uttarakhand   
727  727.0         WB                  West Bengal   
728  728.0         WB                  West Bengal   
729  729.0         WB                  West Bengal   
730  730.0         WB                  West Bengal   
731  731.0         WB                  West Bengal   
732  732.0         WB                  West Bengal   
733  734.0         WB                  West Bengal   
734  735.0         WB                  West Bengal   
735  736.0         WB                  West Bengal   
736  737.0         WB                  West Bengal   
737  738.0         WB                  West Bengal   
738  739.0         WB                  West Bengal   
739  740.0         WB                  West Bengal   
740  741.0         WB                  West Bengal   
741  742.0         WB                  West Bengal   
742  743.0         WB                  West Bengal   
743  744.0         WB                  West Bengal   
744  745.0         WB                  West Bengal   
745  746.0         WB                  West Bengal   
746  747.0         WB                  West Bengal   
747  748.0         WB                  West Bengal   
748  749.0         WB                  West Bengal   
749  750.0         WB                  West Bengal   
750  751.0         WB                  West Bengal   
751  752.0         WB                  West Bengal   
752  753.0         WB                  West Bengal   
753  754.0         WB                  West Bengal   
754  755.0         WB                  West Bengal   

                    District_Key                          Cowin Key  \
0                            NaN                                NaN   
1                    AN_Nicobars                            Nicobar   
2    AN_North and Middle Andaman           North and Middle Andaman   
3               AN_South Andaman                      South Andaman   
4                   AP_Anantapur                          Anantapur   
5                    AP_Chittoor                           Chittoor   
6               AP_East Godavari                      East Godavari   
7                      AP_Guntur                             Guntur   
8                 

In [7]:
#vacc_data.isna().any()

In [8]:
covid_data = pd.read_csv('../data/district_wise.csv')
covid_data

SlNo State_Code                                     State  \
0       0         UN                          State Unassigned   
1       1         AN               Andaman and Nicobar Islands   
2       2         AN               Andaman and Nicobar Islands   
3       3         AN               Andaman and Nicobar Islands   
4       0         AP                            Andhra Pradesh   
5       4         AP                            Andhra Pradesh   
6       5         AP                            Andhra Pradesh   
7       6         AP                            Andhra Pradesh   
8       7         AP                            Andhra Pradesh   
9       8         AP                            Andhra Pradesh   
10      9         AP                            Andhra Pradesh   
11      0         AP                            Andhra Pradesh   
12     10         AP                            Andhra Pradesh   
13     11         AP                            Andhra Pradesh   
14     12         AP                            Andhra Pradesh   
15     13         AP                            Andhra Pradesh   
16     14         AP                            Andhra Pradesh   
17     15         AP                            Andhra Pradesh   
18     16         AP                            Andhra Pradesh   
19     17         AR                         Arunachal Pradesh   
20     18         AR                         Arunachal Pradesh   
21     19         AR                         Arunachal Pradesh   
22     20         AR                         Arunachal Pradesh   
23     21         AR                         Arunachal Pradesh   
24     22         AR                         Arunachal Pradesh   
25     23         AR                         Arunachal Pradesh   
26     24         AR                         Arunachal Pradesh   
27     25         AR                         Arunachal Pradesh   
28     26         AR                         Arunachal Pradesh   
29     27         AR                         Arunachal Pradesh   
..    ...        ...                                       ...   
772   734         CT                              Chhattisgarh   
773   735         DN  Dadra and Nagar Haveli and Daman and Diu   
774   737         DL                                     Delhi   
775   738         GA                                       Goa   
776   739         GJ                                   Gujarat   
777   740         HR                                   Haryana   
778   741         HP                          Himachal Pradesh   
779   742         JK                         Jammu and Kashmir   
780   743         JH                                 Jharkhand   
781   744         KA                                 Karnataka   
782   745         KL                                    Kerala   
783   746         LA                                    Ladakh   
784   747         LD                               Lakshadweep   
785   748         MP                            Madhya Pradesh   
786   749         MH                               Maharashtra   
787   750         MN                                   Manipur   
788   751         ML                                 Meghalaya   
789   752         MZ                                   Mizoram   
790   753         NL                                  Nagaland   
791   754         OR                                    Odisha   
792   755         PY                                Puducherry   
793   756         PB                                    Punjab   
794   757         RJ                                 Rajasthan   
795   758         SK                                    Sikkim   
796   759         TN                                Tamil Nadu   
797   760         TG                                 Telangana   
798   761         TR                                   Tripura   
799   762         UP                             Uttar Pradesh   
800   763         UT                         

In [9]:
df = pd.merge(vacc_data, covid_data, how ='inner', on =['District_Key'])
df

S No State_Code_x                      State_x  \
0      1.0           AN  Andaman and Nicobar Islands   
1      2.0           AN  Andaman and Nicobar Islands   
2      3.0           AN  Andaman and Nicobar Islands   
3      4.0           AP               Andhra Pradesh   
4      5.0           AP               Andhra Pradesh   
5      6.0           AP               Andhra Pradesh   
6      7.0           AP               Andhra Pradesh   
7      8.0           AP               Andhra Pradesh   
8      9.0           AP               Andhra Pradesh   
9     10.0           AP               Andhra Pradesh   
10    11.0           AP               Andhra Pradesh   
11    12.0           AP               Andhra Pradesh   
12    13.0           AP               Andhra Pradesh   
13    14.0           AP               Andhra Pradesh   
14    15.0           AP               Andhra Pradesh   
15    16.0           AP               Andhra Pradesh   
16    17.0           AR            Arunachal Pradesh   
17    18.0           AR            Arunachal Pradesh   
18    19.0           AR            Arunachal Pradesh   
19    20.0           AR            Arunachal Pradesh   
20    21.0           AR            Arunachal Pradesh   
21    22.0           AR            Arunachal Pradesh   
22    34.0           AR            Arunachal Pradesh   
23    23.0           AR            Arunachal Pradesh   
24    24.0           AR            Arunachal Pradesh   
25    25.0           AR            Arunachal Pradesh   
26    26.0           AR            Arunachal Pradesh   
27    27.0           AR            Arunachal Pradesh   
28    28.0           AR            Arunachal Pradesh   
29    29.0           AR            Arunachal Pradesh   
..     ...          ...                          ...   
724  725.0           UT                  Uttarakhand   
725  726.0           UT                  Uttarakhand   
726  727.0           WB                  West Bengal   
727  728.0           WB                  West Bengal   
728  731.0           WB                  West Bengal   
729  729.0           WB                  West Bengal   
730  748.0           WB                  West Bengal   
731  730.0           WB                  West Bengal   
732  752.0           WB                  West Bengal   
733  732.0           WB                  West Bengal   
734  734.0           WB                  West Bengal   
735  735.0           WB                  West Bengal   
736  736.0           WB                  West Bengal   
737  753.0           WB                  West Bengal   
738  737.0           WB                  West Bengal   
739  738.0           WB                  West Bengal   
740  739.0           WB                  West Bengal   
741  740.0           WB                  West Bengal   
742  741.0           WB                  West Bengal   
743  742.0           WB                  West Bengal   
744  743.0           WB                  West Bengal   
745  744.0           WB                  West Bengal   
746  745.0           WB                  West Bengal   
747  746.0           WB                  West Bengal   
748  747.0           WB                  West Bengal   
749  750.0           WB                  West Bengal   
750  749.0           WB                  West Bengal   
751  751.0           WB                  West Bengal   
752  754.0           WB                  West Bengal   
753  755.0           WB                  West Bengal   

                    District_Key                          Cowin Key  \
0                    AN_Nicobars                            Nicobar   
1    AN_North and Middle Andaman           North and Middle Andaman   
2               AN_South Andaman                      South Andaman   
3                   AP_Anantapur                          Anantapur   
4                    AP_Chittoor                           Chittoor   
5               AP_East Godavari                      East Godavari   
6                      AP_Guntur    

In [10]:
#df[df['State_x'] == 'Telangana']
df[df['District_Key'] == 'TG_Bhadradri Kothagudem']
#df[df['State_Code_x'] == 'ML']

S No State_Code_x    State_x             District_Key  \
598  599.0           TG  Telangana  TG_Bhadradri Kothagudem   

                Cowin Key            District_x 16/01/2021 16/01/2021.1  \
598  Bhadradri Kothagudem  Bhadradri Kothagudem       9891            4   

    16/01/2021.2 16/01/2021.3     ...      Active Recovered Deceased  \
598            4          113     ...       18485         4        0   

    Migrated_Other Delta_Confirmed Delta_Active Delta_Recovered  \
598              0               0            0               0   

    Delta_Deceased                                     District_Notes  \
598              0  District-wise numbers are out-dated as cumulat...   

    Last_Updated  
598          NaN  

[1 rows x 2911 columns]

In [11]:
# df[df['State_x'] == 'Gujarat']

In [12]:
# vacc_data[vacc_data['State'] == 'Gujarat']

# Preprocessing

Total 2 Stages are involved in the preprocessing of the data

* **Stage 1**

    * **Step 1** - First I removed extra district code mentioned along with the districts in neighbor.json data i.e District Code followed by '/' --- e.g. converting 'adilabad/Q15211' to 'adilabad
    * **Step 2** - Converted all the districts in title form by removing '_' --- e.g -  'sahibzada_ajit_singh_nagar' to 'Sahibzada Ajit Singh Nagar' 
    * **Step 3** - Some of the districts are writter in the form of 'xxx_district' where xx is district name so,first I converted into title form and then removed the 'district' word from the original name --- e.g. gandhinagar_district to 'Gandhinagar'

In [13]:
print(len(neighbor_data.keys()))
new_neighbor_data={}
i=0
for k, v in neighbor_data.items():
    new_key = k.split('/')[0]
    if new_key in ['bilaspur', 'hamirpur', 'aurangabad', 'pratapgarh', 'balrampur']:
        if new_key in new_neighbor_data.keys():
            key_value=[]
            for val in v:
                val = val.split('/')[0]
                val = val.title().split('_')
                if 'District' in val: val=val[:-1]
                val = " ".join(val)
                key_value.append(val)
            new_neighbor_data[new_key] = key_value
            continue
    
    new_key = new_key.title().split('_')
    if 'District' in new_key:
        #print("before new_key in district", new_key)
        new_key=new_key[:-1]
        #print("after new_key in district", new_key)
    new_key = " ".join(new_key)
    key_value=[]
    for val in v:
        val = val.split('/')[0]
        val = val.title().split('_')
        if 'District' in val: val=val[:-1]
        val = " ".join(val)
        key_value.append(val)
    new_neighbor_data[new_key] = key_value
print("Total input Neighbor-District data available in the datasets are: ", new_neighbor_data)

723
Total input Neighbor-District data available in the datasets are:  {'Rajauri': ['Reasi', 'Jammu', 'Kulgam', 'Punch'], 'Goalpara': ['West Garo Hills', 'North Garo Hills', 'Dhubri', 'Barpeta', 'Bongaigaon', 'Kamrup', 'South Salmara-Mankachar'], 'Mahisagar': ['Dahod', 'Dungarpur', 'Aravalli', 'Panch Mahal', 'Banswara', 'Kheda'], 'Baramula': ['Punch', 'Kupwara', 'Badgam', 'Bandipora', 'Srinagar'], 'Chitradurga': ['Bellary', 'Chikkamagaluru', 'Tumkur', 'Davanagere', 'Anantapur'], 'Khunti': ['Gumla', 'Pashchimi Singhbhum', 'Simdega', 'Seraikela Kharsawan', 'Ranchi'], 'Haridwar': ['Bijnor', 'Dehradun', 'Muzaffarnagar', 'Saharanpur', 'Pauri Garhwal'], 'Baghpat': ['Panipat', 'Shamli', 'Muzaffarnagar', 'Sonipat', 'Ghaziabad', 'North Delhi', 'Meerut'], 'Dima Hasao': ['Cachar', 'West Jaintia Hills', 'East Jaintia Hills', 'West Karbi Anglong', 'Tamenglong', 'Peren', 'Hojai', 'East Karbi Anglong'], 'Mansa': ['Sirsa', 'Bathinda', 'Sangrur', 'Fatehabad', 'Barnala'], 'Dhamtari': ['Kondagaon', 'Gari

In [14]:
# w='Yanam'
# for d, d_n in new_neighbor_data.items():
#     if Levenshtein.distance(w, d) < 3:
#         print(d, end=' ')
#         print(Levenshtein.distance(w, d))
#     for dd in d_n:
#         if Levenshtein.distance(w, dd) < 5:
#             print(d, end=' ')
#             print(Levenshtein.distance(w, dd))

Total input Neighbor-District data available in the datasets are 723

## Note
Here these districts **['bilaspur', 'hamirpur', 'aurangabad', 'pratapgarh', 'balrampur']** are occuring twice in the neighbor-district.json file. So, I combined the neighboring districts of both the repeated district into one.

# Question 1

* **Stage 2**

    * Step 1 - **Finding Intersection of all data** :- Discard all districts which are not available in the all the three datasets i.e Vaccine Data, Covid Data, Neighbor-district Data (Finding intersection of all the datas)
    * Step 2 - **Finding Most Similar Districts** :- Finding the districts from the Neighbor-districts data which are available in intersection of Vaccine Data & Covid data using the technique of finding the similarity score of two words using Levenshtein distance. The Levenshtein distance between two words is basically the minimum number of single-character edits (insertions, deletions or substitutions) required to change one word into the other. (Source - Wiki) 
    * Step 3 - **Manually editing the matched word** :- After getting most similar word, some of the similar word aren't matched properly. So, I tried to find the correct match from internet by manually going through all the incorrect matching.
    * Step 4 - **Correcting spelling mistake** :- Some of the words having wrong spelling written. So I corrected them.

In [15]:
# This function will find the most similar district word from the given district word in the dataset.
def get_most_similar_word(df, w):
    '''
    inputs:
        df : input data containing all the state and districts code i.e DataFrame format
        w : district name
    output:
        output1 : most similar district
        output2 : similarity score between two given word input and most similar word
        
    '''
    temp_df = df.copy()
    temp_df['Levenshtein_dist'] = temp_df['District_x'].apply(lambda x: Levenshtein.distance(w, x))
    temp_df = temp_df.sort_values(by=['Levenshtein_dist'])
    print(temp_df['District_x'].iloc[0])
    #print("with value: ", temp_df['Levenshtein_dist'].iloc[0])
    return temp_df['District_x'].iloc[0], temp_df['Levenshtein_dist'].iloc[0]

In [16]:
def find_correct_match(distr):
    '''
    inputs:
        distr : district name for finding correct match if  Levenshtein_distance doesn't work
    output:
        output1 : correct match if available otherwise return empty string
        
    '''
    correct_match = {'Belgaum': 'Belagavi', 'Baleshwar': 'Balasore', 'Rangareddy': 'Ranga Reddy','Hugli':'Hooghly', 'Sahibzada Ajit Singh Nagar': 'S.A.S. Nagar', 
                 'Sant Ravidas Nagar':'Bhadohi', 'Muktsar': 'Sri Muktsar Sahib', 'Pashchim Champaran':'West Champaran','Pashchimi Singhbhum': 'West Singhbhum',
                'Bid': 'Beed', 'East Karbi Anglong':'Karbi Anglong', 'The Dangs':'Dang', 'Dantewada':'Dakshin Bastar Dantewada','Purbi Singhbhum':'East Singhbhum',
                'Jyotiba Phule Nagar': 'Amroha', 'Kheri':'Lakhimpur Kheri', 'Kaimur (Bhabua)':'Kaimur', 'Ysr':'Y.S.R. Kadapa'}
    if distr in correct_match.keys():
        return correct_match[distr]
    return ''

In [17]:
def find_correct_spelling(distr):
    '''
    inputs:
        distr : district name for correcting speliing mistakes
    output:
        output1 : correct spelling of word if available otherwise return empty string
        
    '''
    correct_spelling = {'Kabeerdham' : 'Kabirdham', 'Shrawasti': 'Shravasti'}
    if distr in correct_spelling.keys():
        return correct_spelling[distr]
    return ''

In [18]:
def is_district_available(distr):
    '''
    inputs:
        distr : district name for checking its availability
    output:
        output1 : True if not available and False if available in the given list
        
    '''
    not_included_in_vacc_and_covid_data = ['Sonapur', 'Faizabad', 'Kochbihar', 'Niwari', 'Noklak']
    if distr in not_included_in_vacc_and_covid_data: return False
    return True

This will take some time to preprocess all the districts and their neighboring districts.

In [ ]:
neighbor_districts_modified={}
for distr, neigh_distr in new_neighbor_data.items():
    if distr in ['Kheri', 'Konkan Division', 'Niwari', 'Noklak', 'Parbhani', 'Pattanamtitta', 
                                    'Chengalpattu', 'Gaurela Pendra Marwahi', 'Nicobars', 'North and Middle Andaman', 
                                    'Seraikela Kharsawan', 'South Andaman', 'Tenkasi', 'Tirupathur', 'Yanam']:
        print('*'*100)
        print("Not including this district: ", distr)
        print('*'*100)
        continue
    if not is_district_available(distr):
        print('*'*100)
        print("District is not available in vaccine and covid data: ", distr)
        print('*'*100)
        continue
    print(distr, end=' ')
    print(' ---> ', end=' ')
    similar_distr, _ = get_most_similar_word(df, distr)
    corr_match = find_correct_match(distr)
    if len(corr_match) > 0: 
        print('*'*100)
        print("Correct Matching of "+distr+" is "+corr_match)
        print('*'*100)
        similar_distr = corr_match
    corr_spelling = find_correct_spelling(similar_distr)
    if len(corr_spelling)>0:
        print('*'*100)
        print("Correct Spelling of "+similar_distr+" is "+corr_spelling)
        print('*'*100)
        
    get_df = df[df['District_x'] == similar_distr]
    district_code = get_df['District_Key'].iloc[0]
    state_code = get_df['State_Code_x'].iloc[0]
    
    i+=1
    distr_code_list=[]
    for n in neigh_distr:
        if n in ['Kheri', 'Konkan Division', 'Niwari', 'Noklak', 'Parbhani', 'Pattanamtitta', 
                                    'Chengalpattu', 'Gaurela Pendra Marwahi', 'Nicobars', 'North and Middle Andaman', 
                                    'Seraikela Kharsawan', 'South Andaman', 'Tenkasi', 'Tirupathur', 'Yanam']:
            print('*'*100)
            print("Not including this district: ", n)
            print('*'*100)
            continue
        if not is_district_available(n): 
            print('*'*100)
            print("District is not available in vaccine and covid data: ", n)
            print('*'*100)
            continue
        print(n, end=' ')
        print(' ---> ', end=' ')
        similar_distr, _ = get_most_similar_word(df, n)
        corr_match = find_correct_match(n)
        if len(corr_match) > 0:
            print('*'*100)
            print("Correct Matching of "+n+" is "+corr_match)
            print('*'*100)
            similar_distr = corr_match
        corr_spelling = find_correct_spelling(similar_distr)
        if len(corr_spelling)>0:
            print('*'*100)
            print("Correct Spelling of "+similar_distr+" is "+corr_spelling)
            print('*'*100)
        n_get_df = df[df['District_x'] == similar_distr]
        i+=1
        n_district_code = n_get_df['District_Key'].iloc[0]
        n_state_code = n_get_df['State_Code_x'].iloc[0]
        distr_code_list.append(n_district_code)
       
    #Handling Last missing data
    if district_code in neighbor_districts_modified.keys():
        print('*'*100)
        print("Combining with previous neighbor district available with same district code: ", district_code)
        print('*'*100)
        distr_code_list = set(neighbor_districts_modified[district_code] + distr_code_list)
    neighbor_districts_modified[district_code] = sorted(distr_code_list)
#neighbor_districts_modified

Rajauri  --->  Rajouri
Reasi  --->  Reasi
Jammu  --->  Jammu
Kulgam  --->  Kulgam
Punch  --->  Punch
Goalpara  --->  Goalpara
West Garo Hills  --->  West Garo Hills
North Garo Hills  --->  North Garo Hills
Dhubri  --->  Dhubri
Barpeta  --->  Barpeta
Bongaigaon  --->  Bongaigaon
Kamrup  --->  Kamrup
South Salmara-Mankachar  --->  South Salmara Mankachar
Mahisagar  --->  Mahisagar
Dahod  --->  Dahod
Dungarpur  --->  Dungarpur
Aravalli  --->  Aravalli
Panch Mahal  --->  Panchmahal
Banswara  --->  Banswara
Kheda  --->  Kheda
Baramula  --->  Baramulla
Punch  --->  Punch
Kupwara  --->  Kupwara
Badgam  --->  Budgam
Bandipora  --->  Bandipora
Srinagar  --->  Srinagar
Chitradurga  --->  Chitradurga
Bellary  --->  Ballari
Chikkamagaluru  --->  Chikkamagaluru
Tumkur  --->  Tumakuru
Davanagere  --->  Davanagere
Anantapur  --->  Anantapur
Khunti  --->  Khunti
Gumla  --->  Gumla
Pashchimi Singhbhum  --->  East Singhbhum
********************************************************************************

Kasganj  --->  Kasganj
Firozabad  --->  Firozabad
Farrukhabad  --->  Farrukhabad
North Goa  --->  North Goa
Sindhudurg  --->  Sindhudurg
South Goa  --->  South Goa
Uttara Kannada  --->  Uttara Kannada
****************************************************************************************************
Not including this district:  Konkan Division
****************************************************************************************************
Belgaum  --->  Budgam
****************************************************************************************************
Correct Matching of Belgaum is Belagavi
****************************************************************************************************
Tiruppur  --->  Tiruppur
Erode  --->  Erode
Karur  --->  Karur
Idukki  --->  Idukki
Dindigul  --->  Dindigul
Coimbatore  --->  Coimbatore
Nadia  --->  Nadia
Murshidabad  --->  Murshidabad
Purba Bardhaman  --->  Purba Bardhaman
Hugli  --->  Puri
******************************************

Pashchimi Singhbhum  --->  East Singhbhum
****************************************************************************************************
Correct Matching of Pashchimi Singhbhum is West Singhbhum
****************************************************************************************************
Gopalganj  --->  Gopalganj
Pashchim Champaran  --->  East Champaran
****************************************************************************************************
Correct Matching of Pashchim Champaran is West Champaran
****************************************************************************************************
Deoria  --->  Deoria
Kushinagar  --->  Kushinagar
Muzaffarpur  --->  Muzaffarpur
Siwan  --->  Siwan
Purba Champaran  --->  East Champaran
Saran  --->  Saran
Seoni  --->  Seoni
Jabalpur  --->  Jabalpur
Narsimhapur  --->  Narsinghpur
Mandla  --->  Mandla
Nagpur  --->  Nagpur
Balaghat  --->  Balaghat
Chhindwara  --->  Chhindwara
Sitapur  --->  Sitapur
Lucknow  --->  Luckn

Kanyakumari  --->  Kanyakumari
Tirunelveli Kattabo  --->  Tirunelveli
Doda  --->  Doda
Anantnag  --->  Anantnag
Udhampur  --->  Udhampur
Chamba  --->  Chamba
Ramban  --->  Ramban
Kishtwar  --->  Kishtwar
Kathua  --->  Kathua
Kallakurichi  --->  Kallakurichi
Perambalur  --->  Perambalur
Tiruvanamalai  --->  Tiruvannamalai
Viluppuram  --->  Viluppuram
Salem  --->  Salem
Cuddalore  --->  Cuddalore
Dharmapuri  --->  Dharmapuri
Thoothukudi  --->  Thoothukkudi
Virudhunagar  --->  Virudhunagar
Ramanathapuram  --->  Ramanathapuram
Tirunelveli Kattabo  --->  Tirunelveli
Kangpokpi  --->  Kangpokpi
Imphal West  --->  Imphal West
Senapati  --->  Senapati
Kamjong  --->  Kamjong
Imphal East  --->  Imphal East
Noney  --->  Noney
Tengnoupal  --->  Tengnoupal
Tamenglong  --->  Tamenglong
Ukhrul  --->  Ukhrul
Bishnupur  --->  Bishnupur
Thoubal  --->  Thoubal
Sahibganj  --->  Sahibganj
Maldah  --->  Malda
Murshidabad  --->  Murshidabad
Katihar  --->  Katihar
Bhagalpur  --->  Bhagalpur
Pakaur  --->  Pakur

Nanded  --->  Nanded
Medak  --->  Medak
Siddipet  --->  Siddipet
Rajgarh  --->  Rajgarh
Guna  --->  Guna
Shajapur  --->  Shajapur
Agar Malwa  --->  Agar Malwa
Sehore  --->  Sehore
Bhopal  --->  Bhopal
Jhalawar  --->  Jhalawar
Rayagada  --->  Rayagada
Koraput  --->  Koraput
Gajapati  --->  Gajapati
Kalahandi  --->  Kalahandi
Kandhamal  --->  Kandhamal
Vizianagaram  --->  Vizianagaram
Srikakulam  --->  Srikakulam
Reasi  --->  Reasi
Udhampur  --->  Udhampur
Punch  --->  Punch
Jammu  --->  Jammu
Ramban  --->  Ramban
Rajauri  --->  Rajouri
Kulgam  --->  Kulgam
West Tripura  --->  West Tripura
Khowai  --->  Khowai
Sepahijala  --->  Sipahijala
Betul  --->  Betul
Harda  --->  Harda
Hoshangabad  --->  Hoshangabad
Khandwa  --->  Khandwa
Amravati  --->  Amravati
Chhindwara  --->  Chhindwara
Aurangabad  --->  Aurangabad
Jalna  --->  Jalna
Nashik  --->  Nashik
Jalgaon  --->  Jalgaon
Ahmednagar  --->  Ahmednagar
Bid  --->  Bidar
***********************************************************************

Rae Bareilly  --->  Rae Bareli
****************************************************************************************************
District is not available in vaccine and covid data:  Faizabad
****************************************************************************************************
Barabanki  --->  Barabanki
Sultanpur  --->  Sultanpur
Chandauli  --->  Chandauli
Varanasi  --->  Varanasi
Ghazipur  --->  Ghazipur
Sonbhadra  --->  Sonbhadra
Kaimur (Bhabua)  --->  Kanpur Nagar
****************************************************************************************************
Correct Matching of Kaimur (Bhabua) is Kaimur
****************************************************************************************************
Mirzapur  --->  Mirzapur
Bargarh  --->  Bargarh
Nuapada  --->  Nuapada
Jharsuguda  --->  Jharsuguda
Sambalpur  --->  Sambalpur
Raigarh  --->  Raigarh
Mahasamund  --->  Mahasamund
Balangir  --->  Balangir
**********************************************************

Kishtwar  --->  Kishtwar
Kargil  --->  Kargil
Kullu  --->  Kullu
Lohit  --->  Lohit
Anjaw  --->  Anjaw
Namsai  --->  Namsai
Lower Dibang Valley  --->  Lower Dibang Valley
Changlang  --->  Changlang
Tinsukia  --->  Tinsukia
Bagalkote  --->  Bagalkote
Raichur  --->  Raichur
Bijapur  --->  Bijapur
Gadag  --->  Gadag
Belgaum  --->  Budgam
****************************************************************************************************
Correct Matching of Belgaum is Belagavi
****************************************************************************************************
Koppal  --->  Koppal
Dungarpur  --->  Dungarpur
Udaipur  --->  Udaipur
Banswara  --->  Banswara
Pratapgarh  --->  Pratapgarh
Mahisagar  --->  Mahisagar
Aravalli  --->  Aravalli
Aligarh  --->  Aligarh
Gautam Buddha Nagar  --->  Gautam Buddha Nagar
Sambhal  --->  Sambhal
Bulandshahr  --->  Bulandshahr
Hathras  --->  Hathras
Kasganj  --->  Kasganj
Mathura  --->  Mathura
Budaun  --->  Budaun
Palwal  --->  Palwal
Banda  --

South Sikkim  --->  South Sikkim
West Sikkim  --->  West Sikkim
Kishanganj  --->  Kishanganj
Jalpaiguri  --->  Jalpaiguri
Mamit  --->  Mamit
Kolasib  --->  Kolasib
North Tripura  --->  North Tripura
Karimganj  --->  Karimganj
Aizwal  --->  Aizawl
Hailakandi  --->  Hailakandi
Lunglei  --->  Lunglei
Buxar  --->  Buxar
Ghazipur  --->  Ghazipur
Bhojpur  --->  Bhojpur
Kaimur (Bhabua)  --->  Kanpur Nagar
****************************************************************************************************
Correct Matching of Kaimur (Bhabua) is Kaimur
****************************************************************************************************
Ballia  --->  Ballia
Rohtas  --->  Rohtas
Gir Somnath  --->  Gir Somnath
Amreli  --->  Amreli
Diu  --->  Diu
Junagadh  --->  Junagadh
Jalandhar  --->  Jalandhar
Kapurthala  --->  Kapurthala
Hoshiarpur  --->  Hoshiarpur
Moga  --->  Moga
Firozpur  --->  Ferozepur
Ludhiana  --->  Ludhiana
Shaheed Bhagat Singh Nagar  --->  Shahid Bhagat Singh Nagar
Kri

Goalpara  --->  Goalpara
East Garo Hills  --->  East Garo Hills
South West Garo Hills  --->  South West Garo Hills
South Salmara-Mankachar  --->  South Salmara Mankachar
South Garo Hills  --->  South Garo Hills
Bhind  --->  Bhind
Etawah  --->  Etawah
Jhansi  --->  Jhansi
Agra  --->  Agra
Datia  --->  Datia
Jalaun  --->  Jalaun
Gwalior  --->  Gwalior
Morena  --->  Morena
Shamli  --->  Shamli
Karnal  --->  Karnal
Panipat  --->  Panipat
Muzaffarnagar  --->  Muzaffarnagar
Saharanpur  --->  Saharanpur
Baghpat  --->  Baghpat
Khagaria  --->  Khagaria
Munger  --->  Munger
Begusarai  --->  Begusarai
Samastipur  --->  Samastipur
Saharsa  --->  Saharsa
Madhepura  --->  Madhepura
Bhagalpur  --->  Bhagalpur
Sitamarhi  --->  Sitamarhi
Sheohar  --->  Sheohar
Muzaffarpur  --->  Muzaffarpur
Darbhanga  --->  Darbhanga
Madhubani  --->  Madhubani
Purba Champaran  --->  East Champaran
Varanasi  --->  Varanasi
Jaunpur  --->  Jaunpur
Ghazipur  --->  Ghazipur
Mirzapur  --->  Mirzapur
Chandauli  --->  Chandaul

Cachar  --->  Cachar
Champhai  --->  Champhai
Noney  --->  Noney
Jiribam  --->  Jiribam
Aizwal  --->  Aizawl
Rewari  --->  Rewari
Gurugram  --->  Gurugram
Nuh  --->  Nuh
Mahendragarh  --->  Mahendragarh
Alwar  --->  Alwar
Jhajjar  --->  Jhajjar
Nawada  --->  Nawada
Sheikhpura  --->  Sheikhpura
Giridih  --->  Giridih
Nalanda  --->  Nalanda
Kodarma  --->  Koderma
Gaya  --->  Gaya
Jamui  --->  Jamui
Longding  --->  Longding
Charaideo  --->  Charaideo
Mon  --->  Mon
Dibrugarh  --->  Dibrugarh
Tirap  --->  Tirap
Datia  --->  Datia
Shivpuri  --->  Shivpuri
Jhansi  --->  Jhansi
Lalitpur  --->  Lalitpur
Bhind  --->  Bhind
Gwalior  --->  Gwalior
Rupnagar  --->  Rupnagar
Hoshiarpur  --->  Hoshiarpur
Una  --->  Una
Solan  --->  Solan
Shaheed Bhagat Singh Nagar  --->  Shahid Bhagat Singh Nagar
Fategarh Sahib  --->  Fatehgarh Sahib
Sahibzada Ajit Singh Nagar  --->  Shahid Bhagat Singh Nagar
****************************************************************************************************
Correct 

Kishtwar  --->  Kishtwar
Anantnag  --->  Anantnag
Lahul And Spiti  --->  Lahaul and Spiti
Chamba  --->  Chamba
Doda  --->  Doda
Kargil  --->  Kargil
Sonitpur  --->  Sonitpur
Darrang  --->  Darrang
Pakke Kessang  --->  Pakke Kessang
West Kameng  --->  West Kameng
Marigaon  --->  Morigaon
Bishwanath  --->  Biswanath
Udalguri  --->  Udalguri
Nagaon  --->  Nagaon
Raichur  --->  Raichur
Bagalkote  --->  Bagalkote
Bellary  --->  Ballari
Jogulamba Gadwal  --->  Jogulamba Gadwal
Bijapur  --->  Bijapur
Kurnool  --->  Kurnool
Koppal  --->  Koppal
Narayanpet  --->  Narayanpet
Yadagiri  --->  Yadgir
Bundi  --->  Bundi
Chittorgarh  --->  Chittorgarh
Bhilwara  --->  Bhilwara
Kota  --->  Kota
Sawai Madhopur  --->  Sawai Madhopur
Tonk  --->  Tonk
Junagadh  --->  Junagadh
Rajkot  --->  Rajkot
Porbandar  --->  Porbandar
Gir Somnath  --->  Gir Somnath
Amreli  --->  Amreli
Barnala  --->  Barnala
Bathinda  --->  Bathinda
Sangrur  --->  Sangrur
Mansa  --->  Mansa
Moga  --->  Moga
Ludhiana  --->  Ludhiana
Ut

Wayanad  --->  Wayanad
Mandya  --->  Mandya
Hassan  --->  Hassan
Hathras  --->  Hathras
Agra  --->  Agra
Aligarh  --->  Aligarh
Kasganj  --->  Kasganj
Mathura  --->  Mathura
Etah  --->  Etah
North 24 Parganas  --->  North 24 Parganas
South 24 Parganas  --->  South 24 Parganas
Nadia  --->  Nadia
Kolkata  --->  Kolkata
Hugli  --->  Puri
****************************************************************************************************
Correct Matching of Hugli is Hooghly
****************************************************************************************************
Howrah  --->  Howrah
East Delhi  --->  East Delhi
Gautam Buddha Nagar  --->  Gautam Buddha Nagar
Central Delhi  --->  Central Delhi
South East Delhi  --->  South East Delhi
Ghaziabad  --->  Ghaziabad
Shahdara  --->  Shahdara
Hingoli  --->  Hingoli
Nanded  --->  Nanded
Washim  --->  Washim
Yavatmal  --->  Yavatmal
****************************************************************************************************
Not incl

Hugli  --->  Puri
****************************************************************************************************
Correct Matching of Hugli is Hooghly
****************************************************************************************************
Puruliya  --->  Purulia
Anuppur  --->  Anuppur
Umaria  --->  Umaria
Mungeli  --->  Mungeli
Dindori  --->  Dindori
Bilaspur  --->  Bilaspur
Koriya  --->  Koriya
Shahdol  --->  Shahdol
Ballia  --->  Ballia
Mau  --->  Mau
Ghazipur  --->  Ghazipur
Deoria  --->  Deoria
Buxar  --->  Buxar
Bhojpur  --->  Bhojpur
Siwan  --->  Siwan
Saran  --->  Saran
Amreli  --->  Amreli
Rajkot  --->  Rajkot
Junagadh  --->  Junagadh
Botad  --->  Botad
Gir Somnath  --->  Gir Somnath
Bhavnagar  --->  Bhavnagar
Kalahandi  --->  Kalahandi
Koraput  --->  Koraput
Nuapada  --->  Nuapada
Kandhamal  --->  Kandhamal
Gariaband  --->  Gariaband
Nabarangapur  --->  Nabarangapur
Rayagada  --->  Rayagada
Balangir  --->  Balangir
Udaipur  --->  Udaipur
Pali  --->  Pali
Rajs

Jayashankar Bhupalapally  --->  Jayashankar Bhupalapally
Bhadradri Kothagudem  --->  Bhadradri Kothagudem
Warangal Rural  --->  Warangal Rural
Bijapur  --->  Bijapur
Mahabubabad  --->  Mahabubabad
Narayanpet  --->  Narayanpet
Kalaburagi  --->  Kalaburagi
Wanaparthy  --->  Wanaparthy
Raichur  --->  Raichur
Jogulamba Gadwal  --->  Jogulamba Gadwal
Vikarabad  --->  Vikarabad
Mahabubnagar  --->  Mahabubnagar
Yadagiri  --->  Yadgir
Tamenglong  --->  Tamenglong
Senapati  --->  Senapati
Cachar  --->  Cachar
Noney  --->  Noney
Kangpokpi  --->  Kangpokpi
Dima Hasao  --->  Dima Hasao
Peren  --->  Peren
Jiribam  --->  Jiribam
Purba Champaran  --->  East Champaran
Pashchim Champaran  --->  East Champaran
****************************************************************************************************
Correct Matching of Pashchim Champaran is West Champaran
****************************************************************************************************
Sheohar  --->  Sheohar
Muzaffarpur  ---

Alappuzha  --->  Alappuzha
Kollam  --->  Kollam
Kottayam  --->  Kottayam
****************************************************************************************************
Not including this district:  Pattanamtitta
****************************************************************************************************
Ernakulam  --->  Ernakulam
Bokaro  --->  Bokaro
Giridih  --->  Giridih
Dhanbad  --->  Dhanbad
Ramgarh  --->  Ramgarh
Hazaribagh  --->  Hazaribagh
Puruliya  --->  Purulia
Pauri Garhwal  --->  Pauri Garhwal
Haridwar  --->  Haridwar
Nainital  --->  Nainital
Bijnor  --->  Bijnor
Tehri Garhwal  --->  Tehri Garhwal
Chamoli  --->  Chamoli
Dehradun  --->  Dehradun
Almora  --->  Almora
Rudraprayag  --->  Rudraprayag
Dimapur  --->  Dimapur
Golaghat  --->  Golaghat
Peren  --->  Peren
Kohima  --->  Kohima
East Karbi Anglong  --->  West Karbi Anglong
****************************************************************************************************
Correct Matching of East Karbi Anglo

Dindigul  --->  Dindigul
Namakkal  --->  Namakkal
Salem  --->  Salem
Chatra  --->  Chatra
Palamu  --->  Palamu
Gaya  --->  Gaya
Latehar  --->  Latehar
Ranchi  --->  Ranchi
Hazaribagh  --->  Hazaribagh
Kottayam  --->  Kottayam
****************************************************************************************************
Not including this district:  Pattanamtitta
****************************************************************************************************
Idukki  --->  Idukki
Alappuzha  --->  Alappuzha
Ernakulam  --->  Ernakulam
Jaisalmer  --->  Jaisalmer
Jodhpur  --->  Jodhpur
Bikaner  --->  Bikaner
Barmer  --->  Barmer
Shajapur  --->  Shajapur
Rajgarh  --->  Rajgarh
Agar Malwa  --->  Agar Malwa
Sehore  --->  Sehore
Ujjain  --->  Ujjain
Dewas  --->  Dewas
Tonk  --->  Tonk
Jaipur  --->  Jaipur
Bundi  --->  Bundi
Ajmer  --->  Ajmer
Bhilwara  --->  Bhilwara
Dausa  --->  Dausa
Sawai Madhopur  --->  Sawai Madhopur
Siddharth Nagar  --->  Siddharthnagar
Balrampur  --->  Balrampur

Meerut  --->  Meerut
Pauri Garhwal  --->  Pauri Garhwal
Udham Singh Nagar  --->  Udham Singh Nagar
Prayagraj  --->  Prayagraj
Jaunpur  --->  Jaunpur
Pratapgarh  --->  Pratapgarh
Mirzapur  --->  Mirzapur
Kaushambi  --->  Kaushambi
Rewa  --->  Rewa
Chitrakoot  --->  Chitrakoot
Sant Ravidas Nagar  --->  Sant Kabir Nagar
****************************************************************************************************
Correct Matching of Sant Ravidas Nagar is Bhadohi
****************************************************************************************************
Sahibzada Ajit Singh Nagar  --->  Shahid Bhagat Singh Nagar
****************************************************************************************************
Correct Matching of Sahibzada Ajit Singh Nagar is S.A.S. Nagar
****************************************************************************************************
Rupnagar  --->  Rupnagar
Ambala  --->  Ambala
Panchkula  --->  Panchkula
Patiala  --->  Patiala
Solan  -

Aligarh  --->  Aligarh
Shahjahanpur  --->  Shahjahanpur
Farrukhabad  --->  Farrukhabad
Papum Pare  --->  Papum Pare
Kamle  --->  Kamle
East Kameng  --->  East Kameng
Kra Daadi  --->  Kra Daadi
Kurung Kumey  --->  Kurung Kumey
Lower Subansiri  --->  Lower Subansiri
Lakhimpur  --->  Lakhimpur
Pakke Kessang  --->  Pakke Kessang
Bishwanath  --->  Biswanath
Lepa Rada  --->  Lepa Rada
Upper Subansiri  --->  Upper Subansiri
Lower Siang  --->  Lower Siang
West Siang  --->  West Siang
Suryapet  --->  Suryapet
Jangaon  --->  Jangaon
Krishna  --->  Krishna
Guntur  --->  Guntur
Nalgonda  --->  Nalgonda
Mahabubabad  --->  Mahabubabad
Khammam  --->  Khammam
Yadadri Bhuvanagiri  --->  Yadadri Bhuvanagiri
Mandla  --->  Mandla
Jabalpur  --->  Jabalpur
Kabirdham  --->  Kabeerdham
****************************************************************************************************
Correct Spelling of Kabeerdham is Kabirdham
**********************************************************************************

Ramban  --->  Ramban
Samba  --->  Samba
Doda  --->  Doda
Kathua  --->  Kathua
Kangra  --->  Kangra
Hoshiarpur  --->  Hoshiarpur
Lahul And Spiti  --->  Lahaul and Spiti
Chamba  --->  Chamba
Una  --->  Una
Gurdaspur  --->  Gurdaspur
Pathankot  --->  Pathankot
Kullu  --->  Kullu
Mandi  --->  Mandi
Hamirpur  --->  Hamirpur
West Sikkim  --->  West Sikkim
Darjeeling  --->  Darjeeling
South Sikkim  --->  South Sikkim
North Sikkim  --->  North Sikkim
Sidhi  --->  Sidhi
Singrauli  --->  Singrauli
Satna  --->  Satna
Rewa  --->  Rewa
Koriya  --->  Koriya
Shahdol  --->  Shahdol
Kollam  --->  Kollam
****************************************************************************************************
Not including this district:  Pattanamtitta
****************************************************************************************************
Alappuzha  --->  Alappuzha
Thiruvananthapuram  --->  Thiruvananthapuram
Tirunelveli Kattabo  --->  Tirunelveli
Kalaburagi  --->  Kalaburagi
Sangareddy  --->  Sa

Gonda  --->  Gonda
Sitapur  --->  Sitapur
Shahjahanpur  --->  Shahjahanpur
Pilibhit  --->  Pilibhit
Bareilly  --->  Bareilly
****************************************************************************************************
Not including this district:  Kheri
****************************************************************************************************
Budaun  --->  Budaun
Farrukhabad  --->  Farrukhabad
Hardoi  --->  Hardoi
Damoh  --->  Damoh
Jabalpur  --->  Jabalpur
Narsimhapur  --->  Narsinghpur
Sagar  --->  Sagar
Chhatarpur  --->  Chhatarpur
Panna  --->  Panna
Katni  --->  Katni
Sonipat  --->  Sonipat
Panipat  --->  Panipat
Jind  --->  Jind
Rohtak  --->  Rohtak
North West Delhi  --->  North West Delhi
North Delhi  --->  North Delhi
Baghpat  --->  Baghpat
Jhajjar  --->  Jhajjar
Vizianagaram  --->  Vizianagaram
Koraput  --->  Koraput
Rayagada  --->  Rayagada
Visakhapatnam  --->  Visakhapatnam
Srikakulam  --->  Srikakulam
Visakhapatnam  --->  Visakhapatnam
Koraput  --->  Korapu

In [ ]:
print("Total Modified Neighbor-district size: ", len(neighbor_districts_modified))

Total Input data is 723 
After preprocessing I got total districts data is 704.
Total 723-704 = 19 sets of data are arranged in this way:
* Out of 19, **4** no. of districts are being removed as per question
* Out of 15, **5** no. of districts are not present in the intersection of vaccine_data and covid data. Refer the **not_included_in_vacc_and_covid_data** list of this function **is_district_available()**
* Remaining out of 10, **5** no. of districts are occuring twice in the neighbor-district.json file. Refer **Note** section of first preprocessing stage.
* Remaining 5 are repeating while combining all neighbor district in the neighbor_districts_modified dict. So, I chose their union of the two sets. Refer in the main code at the end of the for loop.

In [ ]:
sorted_neighbor_districts_modified = dict(sorted(neighbor_districts_modified.items()))
sorted_neighbor_districts_modified

In [ ]:
state_distr_sorted_dict={}
n_state_distr_sorted_dict={}
for state_code in df['State_Code_x'].unique():
    state_distr_sorted_dict[state_code] = []
    
#Changing output to state and district wise code.
for distr, neigh_distr in sorted_neighbor_districts_modified.items():
    get_df = df[df['District_Key'] == distr]
    state_code = get_df['State_Code_x'].iloc[0]
    state_distr_sorted_dict[state_code].append({distr:neigh_distr})
    
#Sorting the districts of states
for st_code, distr_list in state_distr_sorted_dict.items():
    keys_list = []
    for i, k in enumerate(distr_list):
        keys_list.append(list(k.keys())[0])
    sort_keys_list = sorted(keys_list)
    #print(sort_keys_list)
    new_idx_keys_list=[]
    for key in sort_keys_list:
        for kk, vv in enumerate(distr_list):
            if list(vv.keys())[0] == key:
                new_idx_keys_list.append(distr_list[kk])
    if len(new_idx_keys_list) > 0:
        n_state_distr_sorted_dict[st_code] = new_idx_keys_list
n_state_distr_sorted_dict

In [ ]:
print(len(n_state_distr_sorted_dict)) #Two states are not available in their intersections.
print(len(state_distr_sorted_dict)) #Total States 

In [ ]:
json_object = json.dumps(n_state_distr_sorted_dict,sort_keys = True, indent = 4)
with open("../output/neighbor-states-modified.json", "w") as outfile:
    outfile.write(json_object)

In [ ]:
n_state_distr_sorted_dict

In [ ]:
d_json_object = json.dumps(sorted_neighbor_districts_modified, sort_keys = True, indent = 4)
with open("../output/neighbor-districts-modified.json", "w") as outfile:
    outfile.write(d_json_object)

In [ ]:
with open('../output/neighbor-districts-modified.json') as f:
    d = json.load(f)
d